CHAOS-1679 Adds IRSA for AWS (#1878)
harness/developer-hub: docs/chaos-engineering/technical-reference/chaos-faults/aws/aws-iam-integration.md

* **Target accounts:** These are AWS accounts where you'll run experiments against resources and services to intentionally disrupt them. You can have many target accounts.

### Summary of steps for IAM integration

To use IRSA for AWS authentication, you:

1. [Enable the experiment service account to access AWS resources](#enable-the-experiment-service-account-to-access-aws-resources). This enables the Kubernetes service account used to control experiments to access resources across multiple AWS target accounts.

1. [Set up your target AWS accounts for IRSA](#set-up-your-target-accounts-for-irsa). The target accounts are the ones where you will run your chaos experiments.

1. [Establish a trust relationship](#establish-trust-between-the-aws-source-account-and-target-accounts) between the AWS source account and target accounts.

## Enable the experiment service account to access AWS resources

Chaos experiments are initiated and controlled through this service account (usually named `litmus-admin`). You must enable this account to access AWS resources. 

> Follow the steps below to enable the experiment service account to access AWS resources.

Reviewer: ```suggestion
To enable the experiment service account to access AWS resources:
```

> ### Step 1: Create an OIDC provider for your EKS cluster

Reviewer 2: _ignore_

  2021-09-07 14:54:05 [✔]  created IAM Open ID Connect provider for cluster "litmus-demo" in "us-west-1"
  ```

### Step 2: Create an IAM role and policy for your experiment service account 

The experiment service account is usually named `litmus-admin`. You must create an IAM role and policy for this account with the permissions that you would like chaos experiments to have. 

To create the IAM role and policy: 

1. Create a permission policy for the role, and note the policy's Amazon Resource Name (ARN). 

  There are several ways to create a new IAM permission policy. For more information, go to [AWS IAM documentation](https://docs.aws.amazon.com/transfer/latest/userguide/requirements-roles.html). 

  Make a note of the **policy's ARN** for the next step.

1. To create the role and attach the policy you just created, use the `eksctl` command. For example:

  ```bash
  eksctl create iamserviceaccount \
  --name <experiment_service_account_name> \
  --namespace <service_account_namespace> \
  --cluster <cluster_name> \
  --attach-policy-arn <IAM_policy_ARN> \
  --approve \
  --override-existing-serviceaccounts
  ```

  Where:

>     * `experiment_service_account_name` - is the name of the experiment service account, usually `litmus-admin`.

Reviewer: ```suggestion
    * `experiment_service_account_name` is the name of the experiment service account, usually `litmus-admin`.
```

    * `service_account_name` - is the name of the Kubernetes service account, usually `litmus-admin`.
    * `service_account_namespace` is namespace where you installed the execution plane.
    * `IAM_policy_ARN` is the unique Amazon Resource Name (ARN) for the policy you created for this account.

### Step 3: Associate the IAM role with the Kubernetes service account

Associate the IAM role you created in the previous step by annotating the Kubernetes service account (`litmus-admin`). This will give it the required access to AWS resources.

To associate the IAM role to the Kubernetes service account, run this command:

```bash
kubectl annotate serviceaccount -n <SERVICE_ACCOUNT_NAMESPACE> <SERVICE_ACCOUNT_NAME> \
eks.amazonaws.com/role-arn=arn:aws:iam::<ACCOUNT_ID>:role/<IAM_ROLE_NAME>
```

:::note   

> 1. Annotating the `litmus-admin` service account in the `HCE` namespace works for most experiments. 

Reviewer 2: ```suggestion
1. The default name for the experiment service account is litmus-admin, and the namespace for chaos infrastructure is hce. However, users have the flexibility to choose custom values for these settings according to their preferences.
```

> 1. Annotating the `litmus-admin` service account in the `HCE` namespace works for most experiments. 

Reviewer 2: Instead of 

```bash
kubectl annotate serviceaccount -n <SERVICE_ACCOUNT_NAMESPACE> <SERVICE_ACCOUNT_NAME> \
eks.amazonaws.com/role-arn=arn:aws:iam::<ACCOUNT_ID>:role/<IAM_ROLE_NAME>
```
we can say:

```bash
kubectl annotate serviceaccount -n <EXPERIMENT_SERVICE_ACCOUNT_NAMESPACE> <EXPERIMENT_SERVICE_ACCOUNT_NAME> \
eks.amazonaws.com/role-arn=arn:aws:iam::<ACCOUNT_ID>:role/<IAM_ROLE_NAME>
```

1. Provide these required details of the OIDC provider: 
    
    * **Provider URL:** Use the URL you retrieved in Step 1. 
    * **Audience:** Specify `sts.amazonaws.com`.

1. Select **Add provider**.
1. Repeat these steps for each target account.

## Establish trust between the AWS source account and target accounts

The AWS source account enables Harness CE to access resources across multiple target accounts.

### Step 1: Configure trust between target accounts and the AWS source account

You must configure a trust relationship for each IAM role you created in a target account in [Step 1](#step-1-create-an-iam-role-and-policy-in-each-aws-target-account) in the previous section. This authorizes the AWS source account's OIDC provider to assume that IAM role on the target account. To do this, you update the trust policy of the AWS source account, and the policy of the IAM role you created on each target account. 

Follow this procedure for each target account to configure the IAM role and policy on both the AWS source account and the target account.

**To configure trust between the AWS source account OIDC provider and a target account:**

> 
> 1. **In the AWS source account:** Edit the trust relationship in the IAM role YAML file as shown in the example below.

Reviewer 2: ```suggestion

1. **In the AWS source account:** Edit the trust relationship in the IAM role YAML file as shown in the example below.
```

It is not in YAML (instead in JSON) and located in AWS IAM -> role -> trust relationship for all accounts (source and target)

  [[[[ *Still not clear which file they're editing below. We haven't mentioned an IAM role/policy on the source account at this point, so I want to be very specific WHAT they're editing here.* ]]]]

> 

Reviewer 2: There could be a possibility we can select the service under chaos from `source account` then in that case we would need to create IAM role/policy for the source account as well (in the same way how we created for target account).

  In this example `2222222222` is the target account ID, and `1111111111` is the AWS source account ID:

  ```
  {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Principal": {
                  "AWS": "arn:aws:iam::2222222222:role/chaos-role"
              },
              "Action": [
                  "sts:AssumeRole"
              ]
          },
          {
              "Effect": "Allow",
              "Principal": {
                  "Federated": "arn:aws:iam::1111111111:oidc-provider/oidc.eks.us-east-2.amazonaws.com/id/AAAAA11111111C9909AEC6992AEFNWQO0"
              },
              "Action": "sts:AssumeRoleWithWebIdentity"
          }
      ]
  }
  ```

> 2. **In each target account:** Edit the trust relationship in the IAM role YAML file you [created on the target account](#step-1-create-an-iam-role-and-policy-in-each-aws-target-account) as shown in the example below.

Reviewer 2: Same as [this](https://github.com/harness/developer-hub/pull/1878/files#r1218110579)

              "Action": "sts:AssumeRole"
          },
          {
              "Effect": "Allow",
              "Principal": {
                  "Federated": "arn:aws:iam::2222222222:oidc-provider/oidc.eks.us-east-2.amazonaws.com/id/AAAAA11111111C9909AEC6992AEFNWQO0"
              },
              "Action": "sts:AssumeRoleWithWebIdentity"
          }
      ]
  }
  ```

### Step 2: Enable the Kubernetes service account to switch between target accounts

This procedure enables the Kubernetes service account (`litmus-admin`) to seamlessly switch between target accounts when running experiments. To do this you must annotate the Kubernetes service account with the corresponding chaos role in each target account. 

For example, if the target account has a role named `chaos-role`, you must annotate the litmus-admin service account with the unique Amazon Resource Name (ARN) of that role. This enables for seamless switching between target accounts when running experiments.

**To annotate the service account with the role ARN:**

1. Run the following command:

  `kubectl annotate serviceaccount -n <chaos-namespace> <service-account-name> eks.amazonaws.com/role-arn=<role-arn>`

    Where:

>     * `<chaos-namespace>` is the namespace where the service account is installed.

Reviewer 2: ```suggestion
    * `<chaos-namespace>` is the namespace where the chaos infra is installed.
```
(service account is available once the chaos infra is installed)

>     * `<service-account-name>` is the name of your Kubernetes service account.

Reviewer 2: ```suggestion
    * `<service-account-name>` is the name of your Kubernetes experiment service account.
```

```bash
kubectl annotate serviceaccount -n <experiment_service_account_namespace> <experiment_service_account_name> \
eks.amazonaws.com/role-arn=arn:aws:iam::<account_ID>:role/<IAM_role_name>
```

:::note   
1. The default name for the experiment service account is `litmus-admin` and the namespace for chaos infrastructure is `HCE`, however, you can use different names. 
2. For the cluster autoscaler experiment, annotate the experiment service account in the `kube-system` namespace.
:::

### Step 4: Verify the association of the IAM role with the experiment service account

To verify the association between the experiment service account (`litmus-admin`) and the IAM role: 

1. Run an experiment and describe one of the pods. 

1. Verify whether the `AWS_WEB_IDENTITY_TOKEN_FILE` and `AWS_ROLE_ARN` environment variables exist. For example:

  ```bash
  kubectl exec -n litmus <ec2-terminate-by-id-z4zdf> env | grep AWS
  ```
  **Example output:**
  ```
  AWS_VPC_K8S_CNI_LOGLEVEL=DEBUG
  AWS_ROLE_ARN=arn:aws:iam::<ACCOUNT_ID>:role/<IAM_ROLE_NAME>
  AWS_WEB_IDENTITY_TOKEN_FILE=/var/run/secrets/eks.amazonaws.com/serviceaccount/token
  ```

## Set up your target accounts for IRSA

> Whereas chaos experiments are initiated and controlled through the experiment service account, **target accounts** are the accounts you'll subject to chaos experiments so you can intentionally disrupt and manipulate their services.

Reviewer: ```suggestion
Whereas chaos experiments are initiated and controlled through the experiment service account, **target accounts** are the accounts you'll subject to chaos experiments, so you can intentionally disrupt and manipulate their services.
```

In this section, you create an IAM role and set up an OIDC provider in each target account.

### Step 1: Create an IAM role and policy in each AWS target account

This step lets you grant permissions for Harness CE to inject chaos targeting various AWS services in the target account. 

> Create an IAM role and policy in each target account to provide the required permissions for accessing the desired resources in that account. You have the flexibility to define the level of permissions you wish to assign to Harness CE. For instructions, go to [Create an IAM role and policy](https://docs.aws.amazon.com/transfer/latest/userguide/requirements-roles.html) in the AWS documentation.

Reviewer: ```suggestion
Create an IAM role and policy in each target account to provide the required permissions to access the desired resources in that account. You have the flexibility to define the level of permissions you wish to assign to Harness CE. For instructions, go to [Create an IAM role and policy](https://docs.aws.amazon.com/transfer/latest/userguide/requirements-roles.html) in the AWS documentation.
```

1. In the **Add an Identity provider** screen, for **Provider type**, select **OpenID Connect**. 
1. Provide these required details of the OIDC provider: 
    
    * **Provider URL:** Use the URL you retrieved in Step 1. 
    * **Audience:** Specify `sts.amazonaws.com`.

1. Select **Add provider**.
1. Repeat these steps for each target account.

## Establish trust between the AWS source account and target accounts

The AWS source account enables Harness CE to access resources across multiple target accounts.

### Step 1: Configure trust between target accounts and the AWS source account

You must configure a trust relationship for each IAM role you created in a target account in [Step 1](#step-1-create-an-iam-role-and-policy-in-each-aws-target-account) in the previous section. This authorizes the AWS source account's OIDC provider to assume that IAM role on the target account. To do this, you update the trust policy of the AWS source account, and the policy of the IAM role you created on each target account. 

Follow this procedure for each target account to configure the IAM role and policy on both the AWS source account and the target account.

> **To configure trust between the AWS source account OIDC provider and a target account:**

Reviewer: ```suggestion
To configure trust between the AWS source account OIDC provider and a target account:
```

              "Action": "sts:AssumeRoleWithWebIdentity"
          }
      ]
  }
  ```

2. **In each target account:** Edit the trust relationship for the IAM role you [created on the target account](#step-1-create-an-iam-role-and-policy-in-each-aws-target-account) as shown in the example below.

  You can find the JSON for the trust relationship in **AWS IAM > *ROLE_NAME* > Trust relationship** tab.
  
  In this example `2222222222` is the target account ID, and `1111111111` is the experiment service account ID:

  ```
  {
      "Version": "2012-10-17",
      "Statement": [
          {
              "Effect": "Allow",
              "Principal": {
                  "AWS": "arn:aws:iam::1111111111:role/chaos-role"
              },
              "Action": "sts:AssumeRole"
          },
          {
              "Effect": "Allow",
              "Principal": {
                  "Federated": "arn:aws:iam::2222222222:oidc-provider/oidc.eks.us-east-2.amazonaws.com/id/AAAAA11111111C9909AEC6992AEFNWQO0"
              },
              "Action": "sts:AssumeRoleWithWebIdentity"
          }
      ]
  }
  ```

### Step 2: Enable the experiment service account to switch between target accounts

> This procedure enables the experiment service account (`litmus-admin`) to seamlessly switch between target accounts when running experiments. To do this you must annotate the experiment service account with the corresponding chaos role in each target account. 

Reviewer: ```suggestion
This procedure enables the experiment service account (`litmus-admin`) to seamlessly switch between target accounts when running experiments. To do this, you must annotate the experiment service account with the corresponding chaos role in each target account. 
```

> For example, if the target account has a role named `chaos-role`, you must annotate the litmus-admin service account with the unique Amazon Resource Name (ARN) of that role. This enables for seamless switching between target accounts for running experiments.

Reviewer: ```suggestion
For example, if the target account has a role named `chaos-role`, you must annotate the litmus-admin service account with the unique ARN of that role. This enables for seamless switching between target accounts for running experiments.
```
